cni/pkg/iptables/iptables.go

```diff
@@ -0,0 +1,458 @@
+		for _, cmd := range optionalDeleteCmds {
+			err := cfg.ext.Run(iptables, nil, cmd...)
+			if err != nil {
+				log.Debugf("ignoring error deleting optional iptables rule: %v", err)
+			}
+		}
+	}
+	return errors.Join(delErrs...)
+}
+
+// Setup iptables rules for in-pod mode. Ideally this should be an idempotent function.
+// NOTE that this expects to be run from within the pod network namespace!
+func (cfg *IptablesConfigurator) CreateInpodRules(hostProbeSet sets.Set[uint16], hostProbeSNAT *netip.Addr) error {
+	// Append our rules here
+	builder := cfg.appendInpodRules(hostProbeSet, hostProbeSNAT)
+
+	if err := cfg.addLoopbackRoute(); err != nil {
+		return err
+	}
+
+	if err := cfg.addInpodMarkIPRule(); err != nil {
+		return err
+	}
+
+	log.Debug("Adding iptables rules")
+	if err := cfg.executeCommands(builder); err != nil {
+		log.Errorf("failed to restore iptables rules: %v", err)
+		return err
+	}
+
+	return nil
+}
+
+func (cfg *IptablesConfigurator) appendInpodRules(hostProbeSet sets.Set[uint16], hostProbeSNAT *netip.Addr) *builder.IptablesBuilder {
+	inpodMark := fmt.Sprintf("0x%x", InpodMark) + "/" + fmt.Sprintf("0x%x", InpodMask)
+	inpodTproxyMark := fmt.Sprintf("0x%x", InpodTProxyMark) + "/" + fmt.Sprintf("0x%x", InpodTProxyMask)
+
+	iptablesBuilder := builder.NewIptablesBuilder(ipbuildConfig(cfg.cfg))
+
+	// Insert jumps to our custom chains
+
+	// -t mangle -A OUTPUT -j ISTIO_PRERT
+	iptablesBuilder.AppendRule(
+		iptableslog.UndefinedCommand, iptablesconstants.PREROUTING, iptablesconstants.MANGLE,
+		"-j", ChainInpodPrerouting,
+	)
+
+	// -t mangle -A OUTPUT -p tcp -j ISTIO_OUTPUT
+	iptablesBuilder.AppendRule(
+		iptableslog.UndefinedCommand, iptablesconstants.OUTPUT, iptablesconstants.MANGLE,
+		"-j", ChainInpodOutput,
+	)
+
+	// -t nat -A OUTPUT -p tcp -j ISTIO_OUTPUT
+	iptablesBuilder.AppendRule(
+		iptableslog.UndefinedCommand, iptablesconstants.OUTPUT, iptablesconstants.NAT,
+		"-j", ChainInpodOutput,
+	)
+
+	// From here, we should be only inserting rules into our custom chains.
+
+	// -A PREROUTING -m mark --mark 0x3/0xfff -j CONNMARK --set-xmark 0x111/0xfff
+	iptablesBuilder.AppendRule(iptableslog.UndefinedCommand, ChainInpodPrerouting, iptablesconstants.MANGLE, "-m", "mark",
+		"--mark", inpodMark,
+		"-j", "CONNMARK",
+		"--set-xmark", inpodTproxyMark)
+
+	// Handle healthcheck probes from the host node. In the host netns, before the packet enters the pod, we SNAT
+	// the healthcheck packet to a fixed IP if the packet is coming from a node-local process with a socket.
+	//
+	// We do this so we can exempt this traffic from ztunnel capture/proxy - otherwise both kube-proxy (legit)
+	// and kubelet (skippable) traffic would have the same srcip once they got to the pod, and would be indistinguishable.
+	//
+	// Note that SortedList is used here because the DIY istio sets class has no order guarantees,
+	// and our unit tests will flake if rules have a nondeterministic ordering :D. Shoulda used slices.
```

Comment:
I'll amend the comment, I was being needlessly fussy here. `set` makes sense here since we are relying on it to ensure a unique set of ports earlier in the code, and anyway sorting the list here is not a concern.